Koha/Illrequest.pm

```diff
@@ -1346,57 +1330,282 @@ EOF
-                message => $Mail::Sendmail::error,
-                method  => 'generic_confirm',
-                stage   => 'draft',
-            };
         }
+        return {
+            error   => 1,
+            status  => 'email_failed',
+            message => 'Email queueing failed',
+            method  => 'generic_confirm',
+            stage   => 'draft',
+        };
     } else {
         die "Unknown stage, should not have happened."
     }
 }
 
+=head3 get_staff_to_address
+
+    my $email = $request->get_staff_to_address();
+
+Get the email address to which staff notices should be sent
+
+=cut
+
+sub get_staff_to_address {
+    my ( $self ) = @_;
+
+    # The various places we can get an ILL staff email address from
+    # (In order of preference)
+    #
+    # Dedicated branch address
+    my $library = Koha::Libraries->find( $self->branchcode );
+    my $branch_ill_to = $library->branchillemail;
+    # General purpose ILL address from syspref
+    my $syspref = C4::Context->preference("ILLDefaultStaffEmail");
+    # Branch general email address
+    my $branch_to = $library->branchemail;
+    # Last resort
+    my $koha_admin = C4::Context->preference('KohaAdminEmailAddress');
+
+    my $to;
+    if ($branch_ill_to) {
+        $to = $branch_ill_to;
+    } elsif ($syspref) {
+        $to = $syspref;
+    } elsif ($branch_to) {
+        $to = $branch_to;
+    } elsif ($koha_admin) {
+        $to = $koha_admin;
+    }
+
+    # $to will not be defined if we didn't find a usable address
+    return $to;
+}
+
+=head3 send_patron_notice
+
+    my $result = $request->send_patron_notice($notice_code);
+
+Send a specified notice regarding this request to a patron
+
+=cut
+
+sub send_patron_notice {
+    my ( $self, $notice_code ) = @_;
+
+    # We need a notice code
+    if (!$notice_code) {
+        return {
+            error => 'notice_no_type'
+        };
+    }
+
+    # Map from the notice code to the messaging preference
+    my %message_name = (
+        ILL_PICKUP_READY   => 'Ill_ready',
+        ILL_REQUEST_UNAVAIL => 'Ill_unavailable'
+    );
+
+    # Get the patron's messaging preferences
+    my $borrower_preferences = C4::Members::Messaging::GetMessagingPreferences({
+        borrowernumber => $self->borrowernumber,
+        message_name   => $message_name{$notice_code}
+    });
+    my @transports = keys %{ $borrower_preferences->{transports} };
+
+    # Send the notice to the patron via the chosen transport methods
+    # and record the results
+    my @success = ();
+    my @fail = ();
+    for my $transport (@transports) {
+        my $letter = $self->get_notice({
+            notice_code => $notice_code,
+            transport   => $transport
+        });
+        if ($letter) {
+            my $result = C4::Letters::EnqueueLetter({
+                letter                 => $letter,
+                borrowernumber         => $self->borrowernumber,
+                message_transport_type => $transport,
```

Comment:
Had we not auth to set 'from_address' and 'reply_address' here?

Reply:
from_address will be appropriately set at send time, reply_address is less clear... should patrons ever be expected to want to reply to such a notice and if so would one expect the reply to go to the branchreplyto or the branchillemail